### PR TITLE
[cli][fs-detectors] Improve `vc build` monorepo support

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
### Related Issues

Improves how `vc build` handles monorepos. In short, this introduces monorepo manager detection logic and then some helpful defaults so users don't have to manually specify a `buildCommand` or `installCommand` when linking a project within a monorepo.

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/cli/src/commands/build.ts [hunk 1]
```diff
@@ -271,6 +276,12 @@ async function doBuild(
   outputDir: string
 ): Promise<void> {
   const { output } = client;
+
+  const framework = await detectFramework({
```
Comment (agadzik) [inline packages/cli/src/commands/build.ts]:
```suggestion
  const monorepoManager = await detectFramework({
```

File: packages/cli/src/commands/build.ts [hunk 2]
```diff
@@ -432,6 +443,28 @@ async function doBuild(
...
+          join(cwd, projectDirectory),
+          cwd
+        )} && turbo run build --filter=${basename(projectDirectory)}...`,
+      },
+    };
+    if (framework === 'turbo') {
+      return monorepoDefaults.turbo.buildCommand;
+    }
+
+    return undefined;
```
Comment (agadzik) [inline packages/cli/src/commands/build.ts]:
```suggestion
    return monorepoDefaults[monorepoManager]?.buildCommand;
```

File: packages/fs-detectors/src/packageManagers/package-managers.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    detectors: {
+      some: [
+        {
+          path: 'yarn.lock',
+        },
+        {
+          path: 'package.json',
+          matchContent: '"packageManager":\\s*"yarn@.*"',
+        },
+      ],
```
Comment (agadzik) [inline packages/fs-detectors/src/packageManagers/package-managers.ts]: you might want to add another rule here for when a repo doesnt have a lock file, and has a package file with no `packageManager` field set. 

```ts
{
  path: 'package.json'
}
```

(we default to yarn in this case)

File: packages/cli/src/commands/build.ts [hunk 4]
```diff
@@ -301,6 +308,145 @@ async function doBuild(
     ...pickOverrides(localConfig),
   };
 
+  const localFileSystem = new LocalFileSystemDetector(cwd);
+
+  const framework = await detectFramework({
```
Comment (agadzik) [inline packages/cli/src/commands/build.ts]: the name `framework` is is misleading, can we rename to `monorepoManager` instead?

File: packages/cli/package.json [hunk 5]
```diff
@@ -81,7 +82,7 @@
     "@types/minimatch": "3.0.3",
     "@types/mri": "1.1.0",
     "@types/ms": "0.7.30",
-    "@types/node": "11.11.0",
+    "@types/node": "16.11.7",
```
Comment (Ethan-Arrowood) [inline packages/cli/package.json]: This version update introduced some type changes throughout the CLI. Might be worth doing this change in a separate PR.

File: packages/cli/package.json [hunk 6]
```diff
@@ -14,8 +14,9 @@
     "preinstall": "node ./scripts/preinstall.js",
     "test": "jest --env node --verbose --runInBand --bail --forceExit",
     "test-unit": "yarn test test/unit/",
-    "test-integration-cli": "rimraf test/fixtures/integration && ava test/integration.js --serial --fail-fast --verbose",
+    "test-integration-cli": "rimraf test/fixtures/integration && ava test/integration.js --serial --fail-fast --verbose && yarn test:integration",
     "test-integration-dev": "yarn test test/dev/",
+    "test:integration": "jest --env node --verbose --bail test/integration/",
```
Comment (Ethan-Arrowood) [inline packages/cli/package.json]: We've discussed improving the tests in this repo for a while. I think now is as good a time as any. This change only affects the new tests and uses Jest (which is already configured for the unit tests). 

I'll extrapolate on this change in our slack channel, but in short, by using a test framework that supports test scoping with beforeEach/afterEach blocks allows the integration tests for this feature to be a lot cleaner.

File: packages/cli/src/commands/build.ts [hunk 7]
```diff
@@ -57,6 +63,7 @@ import { initCorepack, cleanupCorepack } from '../util/build/corepack';
 import { sortBuilders } from '../util/build/sort-builders';
 import { toEnumerableError } from '../util/error';
 import { validateConfig } from '../util/validate-config';
+import { Output } from '../util/output';
```
Comment (Ethan-Arrowood) [inline packages/cli/src/commands/build.ts]:
```suggestion
import type { Output } from '../util/output';
```

File: packages/cli/src/util/client.ts [hunk 8]
```diff
@@ -93,7 +92,7 @@ export default class Client extends EventEmitter implements Stdio {
       : '';
 
     if (opts.accountId || opts.useCurrentTeam !== false) {
-      const query = new URLSearchParams(parsedUrl.query);
+      const query = new URLSearchParams(parsedUrl.query as any);
```
Comment (Ethan-Arrowood) [inline packages/cli/src/util/client.ts]: changes in this file are related to the `@types/node` version update

File: packages/cli/src/util/dev/builder.ts [hunk 9]
```diff
@@ -87,7 +87,7 @@ async function createBuildProcess(
 
   return new Promise((resolve, reject) => {
     // The first message that the builder process sends is the `ready` event
-    buildProcess.once('message', ({ type }) => {
+    buildProcess.once('message', ({ type }: { type: string }) => {
```
Comment (Ethan-Arrowood) [inline packages/cli/src/util/dev/builder.ts]: changes in this file are related to the `@types/node` version update

File: packages/cli/src/util/dev/parse-listen.ts [hunk 10]
```diff
@@ -31,11 +31,11 @@ export function parseListen(str: string, defaultPort = 3000): ListenSpec {
       return [url.pathname];
     case 'tcp:':
       url.port = url.port || String(defaultPort);
-      return [parseInt(url.port, 10), url.hostname];
+      return [parseInt(url.port, 10), url.hostname as string];
```
Comment (Ethan-Arrowood) [inline packages/cli/src/util/dev/parse-listen.ts]: changes in this file are related to the `@types/node` version update

File: packages/cli/src/util/dev/router.ts [hunk 11]
```diff
@@ -58,7 +58,7 @@ export async function devRouter(
 ): Promise<RouteResult> {
   let result: RouteResult | undefined;
   let { pathname: reqPathname = '/', search: reqSearch } = url.parse(reqUrl);
-  const reqQuery = parseQueryString(reqSearch);
+  const reqQuery = parseQueryString(reqSearch as string);
```
Comment (Ethan-Arrowood) [inline packages/cli/src/util/dev/router.ts]: changes in this file are related to the `@types/node` version update

File: packages/cli/src/util/dev/server.ts [hunk 12]
```diff
@@ -996,7 +996,7 @@ export default class DevServer {
     }
 
     if (devProcess) {
-      ops.push(treeKill(devProcess.pid));
+      ops.push(treeKill(devProcess.pid as number));
```
Comment (Ethan-Arrowood) [inline packages/cli/src/util/dev/server.ts]: changes in this file are related to the `@types/node` version update

File: packages/cli/test/helpers/prepare.js [hunk 13]
```diff
@@ -19,7 +19,8 @@ const getRevertAliasConfigFile = () => {
     ],
   });
 };
-module.exports = async function prepare(session, binaryPath, tmpFixturesDir) {
+
+async function prepareFixtures(session, binaryPath, tmpFixturesDir) {
```
Comment (Ethan-Arrowood) [inline packages/cli/test/helpers/prepare.js]: this function matches the previous default export prepare function from this file. it is used in `integration.js` to keep it operating as usual

File: packages/cli/test/helpers/prepare.js [hunk 14]
```diff
@@ -536,15 +537,454 @@ module.exports = async function prepare(session, binaryPath, tmpFixturesDir) {
...
   };
 
-  for (const [typeName, needed] of Object.entries(spec)) {
-    const directory = join(tmpFixturesDir, typeName);
+  for (const [fixtureName, fixtureContents] of Object.entries(spec)) {
+    prepare(tmpFixturesDir, fixtureName, fixtureContents);
+  }
+}
+
+const fixtures = {
```
Comment (Ethan-Arrowood) [inline packages/cli/test/helpers/prepare.js]: this list of fixtures is all new. follows same format as previous fixture list so that it can be used with the same `prepare` function logic

Comment (Ethan-Arrowood) [reply packages/cli/test/helpers/prepare.js]: I'm open to the idea of changing this data structure to something that allows more metadata maybe. Something along the lines of:

```js
const fixtures = [
  {
    name: 'monorepo-detection-turbo',
    content: [
      { name: '.vercel/project.json', data: JSON.stringify({ /* ... */ }), },
      /* ... */
    ],
  },
]
```

File: packages/fs-detectors/src/isError.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
+// These functions are copied from vercel/cli
```
Comment (Ethan-Arrowood) [inline packages/fs-detectors/src/isError.ts]: I'd love to move these functions to some shared directory within the monorepo. Then they can be tested and used reliably by any package within the repo. Can be a private package too

File: packages/fs-detectors/src/packageManagers/package-managers.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
+import type { Framework } from '@vercel/frameworks';
+
+export const packageManagers: Array<
+  Omit<Framework, 'description' | 'getOutputDirName' | 'settings'>
+> = [
```
Comment (Ethan-Arrowood) [inline packages/fs-detectors/src/packageManagers/package-managers.ts]: Now that we are using the concept of a _framework_ to represent monorepo managers, workspace managers, and package managers, we may want to rework some of this. Maybe at least the naming of things. It shouldn't be `frameworkDetector` but maybe just `entityDetector`, and it shouldn't require (but can still support) all these properties that were added just for frameworks

File: packages/fs-detectors/test/unit.local-file-system-detector.test.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
+import os from 'node:os';
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import { LocalFileSystemDetector, DetectorFilesystem } from '../src';
+
+const tmpdir = path.join(os.tmpdir(), 'local-file-system-test');
+
+const dirs = ['', 'a', 'a/b']; // root, single-nested, double-nested
+const files = ['foo', 'bar'];
+const filePaths = dirs.flatMap(dir => files.map(file => path.join(dir, file)));
```
Comment (Ethan-Arrowood) [inline packages/fs-detectors/test/unit.local-file-system-detector.test.ts:10]: This file structure should cover all the relevant bases. root files, single nested files, and multie nested files.

**Review by Ethan-Arrowood — COMMENTED** [review]
Self-reviewed to get discussions going. This PR has become very large, I think I can pull out the `@types/node` update if folks care about scope-creep. 

I'm going to say that the integration test changes are necessary. The existing integration test file was simply not extensible enough to support the multitude of scenarios we need to be testing for this feature.

File: packages/fs-detectors/src/isError.ts (hunk 15 above)
Comment (TooTallNate) [reply packages/fs-detectors/src/isError.ts]: Maybe we make a `@vercel/error` package, so it can be used elsewhere as well?

File: packages/cli/package.json (hunk 6 above)
Comment (TooTallNate) [reply packages/cli/package.json]: I'm curious why these couldn't just be unit tests

Comment (Ethan-Arrowood) [reply packages/cli/package.json]: I believe our reasoning was that they are invoking the CLI directly. While yes, they are all just using `vc build` instead of `vc deploy` or something. The fact that it was using the compiled CLI made us put them in the integration bucket. 

I guess we could rewrite them to call `build` directly instead of invoking through CLI

Comment (Ethan-Arrowood) [reply packages/cli/package.json]: The CI has been really flaky with the integration tests I added because of how long they take to run sometimes. 

If it is sufficient, I'm in favor of moving them to unit tests to be more efficient.

File: packages/cli/src/commands/build.ts [hunk 18]
```diff
@@ -601,6 +611,134 @@ async function doBuild(
   );
 }
 
+async function setMonorepoDefaultSettings(
```
Comment (TooTallNate) [inline packages/cli/src/commands/build.ts]: Let's move this to a different file in `src/util/commands/build`?

**Review by TooTallNate — APPROVED** [review]
<picture data-single-emoji=":rockout:" title=":rockout:">:rockout:</picture>

**Comment by Ethan-Arrowood** [comment]
<picture data-single-emoji=":dancing_jason_and_freddy:" title=":dancing_jason_and_freddy:">:dancing_jason_and_freddy:</picture>
